src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -76,17 +95,31 @@ internal class EventBasedLRUCache<TKey, TValue> : IDisposable
         {
             _tryTakeTimeout = tryTakeTimeout;
             _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _options = options;
             _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _cleanUpIntervalInMilliSeconds = 1000 * cleanUpIntervalInSeconds;
             _removeExpiredValues = removeExpiredValues;
-            _eventQueueTask = new Task(OnStart, options);
-            _eventQueueTask.Start();
+
             if (_removeExpiredValues)
-                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+                _timer = new Timer(RemoveExpiredValuesPeriodically, null, _cleanUpIntervalInMilliSeconds, _cleanUpIntervalInMilliSeconds); // initial delay then ticks every periodInMilliSeconds
         }
 
-        private void OnStart()
+        /// <summary>
+        /// This is the delegate for starting the event queue task (and the timer to remove the expired items if _removeExpiredValues is true).
+        /// The task keeps running until it is disposed or cancelled (_eventQueueTaskState is set to EventQueueTaskStopRequested).
+        /// The task and timer are synchronized; both are running or stopped.
+        /// </summary>
+        private void EventQueueTaskAction()
         {
-            while (!_disposed)
+            Interlocked.Increment(ref _taskCount);
+
+            if (_removeExpiredValues)
+            {
+                ResumeTimer();
+            }
+
+            // keep running until it is to be disposed, or when asked to stop (EventQueueTaskStopping)
+            while (!_disposed && Interlocked.CompareExchange(ref _eventQueueTaskState, EventQueueTaskStopped, EventQueueTaskStopRequested) != EventQueueTaskStopRequested)
```

Comment:
Can you explain why this is being compared to EventQueueTaskStopRequested (rather than EventQueueTaskStopped)?

Reply:
1. If the _eventQueueTaskState is EventQueueTaskRunning, the comparison is true so the while loop continues.
2. If the _eventQueueTaskState is EventQueueTaskStopRequested (a stop was requested via the OnLinkedListItemRemoved when a cache item was removed) then the check _eventQueueTaskState != _eventQueueTaskState is false and the task exits the while loop and the _eventQueueTaskState is set to EventQueueTaskStopped.